src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -7130,24 +7144,188 @@ public override IOperation VisitArgument(IArgumentOperation operation, int? capt
+                var outerCaptureRegion = CurrentRegionRequired;
+
+                // The inner region holds the capture for the operand (ie. old value)
+                var innerCaptureRegion = new RegionBuilder(ControlFlowRegionKind.LocalLifetime);
+                EnterRegion(innerCaptureRegion);
+
+                var initializers = operation.Initializer.Initializers;
+
+                var properties = operation.Type.GetMembers()
+                    .Where(m => m.Kind == SymbolKind.Property)
+                    .SelectAsArray(m => (IPropertySymbol)m);
+
+                int oldValueCaptureId;
+                if (setAllProperties(initializers, properties))
+                {
+                    // Avoid capturing the old value since we won't need it
+                    oldValueCaptureId = -1;
+                    AddStatement(VisitRequired(operation.Operand));
+                }
+                else
+                {
+                    oldValueCaptureId = GetNextCaptureId(innerCaptureRegion);
+                    VisitAndCapture(operation.Operand, oldValueCaptureId);
+                }
+                // calls to Visit may enter regions, so we reset things
+                LeaveRegionsUpTo(innerCaptureRegion);
+
+                var explicitProperties = PooledDictionary<string, IOperation>.GetInstance();
+                var initializerBuilder = ArrayBuilder<IOperation>.GetInstance(initializers.Length);
+
+                // Visit and capture all the values, and construct assignments using capture references
+                foreach (IOperation initializer in initializers)
+                {
+                    var simpleAssignment = (ISimpleAssignmentOperation)initializer;
+
+                    int valueCaptureId = GetNextCaptureId(outerCaptureRegion);
+                    VisitAndCapture(simpleAssignment.Value, valueCaptureId);
+                    LeaveRegionsUpTo(innerCaptureRegion);
+                    var valueCaptureRef = new FlowCaptureReferenceOperation(valueCaptureId, operation.Operand.Syntax,
+                        operation.Operand.Type, constantValue: operation.Operand.GetConstantValue());
+
+                    if (simpleAssignment.Target is InvalidOperation)
+                    {
+                        var badTarget = new InvalidOperation(ImmutableArray<IOperation>.Empty, semanticModel: null, simpleAssignment.Target.Syntax,
+                            type: null, constantValue: null, isImplicit: true);
+
+                        var badAssignment = new SimpleAssignmentOperation(isRef: false, badTarget, valueCaptureRef,
+                            semanticModel: null, operation.Syntax, valueCaptureRef.Type, constantValue: null, isImplicit: true);
+
+                        initializerBuilder.Add(badAssignment);
+                        continue;
+                    }
+
+                    var propertyReference = (IPropertyReferenceOperation)simpleAssignment.Target;
```

Comment:
>(IPropertyReferenceOperation)simpleAssignment.Target; [](http://example.com/codeflow?start=44&length=53)

Is this cast always safe?

Reply:
I think so. Nested initializers and indexing initializers are parsing errors. Expressions fall into the InvalidOperation check above.